### ISSUE TITLE
Wrong `since` frontier for merged batch?

### ISSUE THREAD

**Issue by LorenzSelv** [TimelyDataflow/differential-dataflow#247]
https://github.com/TimelyDataflow/differential-dataflow/blob/e50c9ce938c9783d7a5eb7fdf25f047f2c6ed145/src/trace/implementations/ord.rs#L227

Consider the full set of updates of e.g. a simple wordcount dataflow (single-dimensional timestamps):
```
full = { ("hello", (), t=5, +1) }
```

Suppose we seal two batches: `batch1` contains the only update, `batch2` is empty.
```
batch1 = Description { lower: 0,  upper: 10, since: 0 }
         Content { ("hello", (), t=5, +1) }
batch2 = Description { lower: 10, upper: 20, since: 10 }
         Content { }
```

Suppose the times inside `batch1` are advanced to timestamp `15`
```
batch1 = Description { lower: 0,  upper: 10, since: 15 }
         Content { ("hello", (), t=15, +1) }  // times in this batch has been advanced to t=15
batch2 = Description { lower: 10, upper: 20, since: 10 }
         Content { }
```

Suppose we merge the two batches. According to the code, the resulting batch would have the following description / content
```
Description { lower: 0, upper: 20, since: 10 }
Content { ("hello", (), t=15, +1) }
```
The `since` frontier is taken to be the **minimum** of the two `since` frontiers of the batches, so `batch2.since = 10` is taken.

However, the [property of the batch](https://github.com/TimelyDataflow/differential-dataflow/blob/56a7eeddbbccce46c2132620c6dbb2f0bff7bcc7/src/trace/description.rs#L33) does not hold anymore.

Consider for example `t=12`, for which it should hold as it is larger than the `since` frontier.
There are no accumulated updates before such time, which is not the case if we consider the
original set of updates (`full`).

To conclude, I think the `since` frontier of the merged batch should be the **maximum** of the two original batches (`t=15` in this case). And it's easy to see that the property would hold for the case above.

The intuition that justifies this is that if a batch looses the capability to distinguish since a certain timestamp (as `batch1` did) also a resulting merged batch will not have such capability.

-------

On a similar note, I believe there is another issue in the `work` function that performs the merging.
https://github.com/TimelyDataflow/differential-dataflow/blob/e50c9ce938c9783d7a5eb7fdf25f047f2c6ed145/src/trace/implementations/ord.rs#L293

When a frontier (call it `compaction_frontier`) is supplied, times are advanced to such frontier and compaction of updates is performed. If the `compaction_frontier` is larger than the `since` frontier that was determined in the constructor, then the `since` frontier should be updated to be the `compaction_frontier`. That is because again we are not able to fulfill the property above (we are loosing the ability to distinguish since that frontier).

So the following snippet should be updated to
```rust
// if we are supplied a frontier, we should compact.
if let Some(frontier) = frontier.as_ref() {
    OrdValBatch::advance_builder_from(&mut self.result, frontier, initial_key_pos);
    if self.description.since().iter().all(|t1| frontier.iter().any(|t2| t1.less_equal(t2))) {
        self.description = Description::new(self.description.lower(), self.description.upper(), frontier);
    }
}
```

**Comment by frankmcsherry** [comment]
It's definitely the case that `since` logic has not had a lot of love. It's in place because the concept is important, but I'm not sure it's ever read (and consequently, probably never debugged either).

The since frontier should be the "max" of the two since frontiers, as you correctly note. When we seal a batch, we should be able to use `[0]` as the `since` frontier (it's only once compaction is applied that we need to be careful). And, when compaction happens we certainly need to advance the since frontier.

I'll take a look at this, or if you'd like to submit a PR that would be totally welcome too!

**Comment by LorenzSelv** [comment]
I'm using the `since` logic as part of the fault tolerance project I'm working on. I discovered this as the state I was reconstructing after failure was not consistent what it was supposed to be.

Maybe double check what I wrote above makes sense to you then I'd be happy to submit the PR.

**Comment by frankmcsherry** [comment]
Yup. Everything you've said makes sense.

**Comment by frankmcsherry** [comment]
Also worth checking out the batch formation, and that `since` of `vec![T::minimum()]` gets used there. I've looked a few places and none do the right thing so far.

**Comment by frankmcsherry** [comment]
Actually, one minor correction, maybe not actually a problem in practice but:

When we have two `since` frontiers, perhaps from two batches or one batch and a compaction frontier, we aren't necessarily sure that one is "larger" than the other. Instead, we want the frontier that describes times beyond both frontiers.

This isn't that complicated. It should be
```rust
let mut upper = Antichain::new();
for time1 in since_1 {
    for time2 in since_2 {
        upper.insert(time1.join(time2));
    }
}
```
This introduces pairs of joins, which ensures that it captures exactly those forward cones that are the join of each pair of forward cones from each since frontier.
